crates/core/src/subscription/module_subscription_actor.rs

```diff
@@ -51,38 +57,50 @@ impl ModuleSubscriptions {
             queries.extend(qset);
         }
 
-        let subscription = match self.subscriptions.iter_mut().find(|s| s.queries == queries) {
+        let database_update = tokio::task::block_in_place(|| queries.eval(&self.relational_db, tx, auth))?;
+        Ok((queries, database_update))
+    }
+
+    /// Add a subscriber to the module. NOTE: this function is blocking.
+    pub fn add_subscriber(&self, sender: ClientConnectionSender, subscription: Subscribe) -> Result<(), DBError> {
+        let (queries, database_update) = self.db_updates(&sender, subscription)?;
+        // holds lock after `eval` so that it can be concurrent
+        let mut subscriptions = self.subscriptions.write();
+        self._remove_subscriber(sender.id, &mut subscriptions);
+        let subscription = match subscriptions.iter_mut().find(|s| s.queries == queries) {
             Some(sub) => {
                 sub.add_subscriber(sender);
                 sub
             }
             None => {
                 let n = queries.len();
-                self.subscriptions.push(Subscription::new(queries, sender));
+                subscriptions.push(Subscription::new(queries, sender));
                 WORKER_METRICS
                     .subscription_queries
                     .with_label_values(&self.relational_db.address())
                     .add(n as i64);
-                self.subscriptions.last_mut().unwrap()
+                subscriptions.last_mut().unwrap()
             }
         };
 
-        let database_update = tokio::task::block_in_place(|| subscription.queries.eval(&self.relational_db, tx, auth))?;
-
-        let sender = subscription.subscribers().last().unwrap();
-
+        let sender = subscription.subscribers().last().unwrap().clone();
+        drop(subscriptions);
         // NOTE: It is important to send the state in this thread because if you spawn a new
         // thread it's possible for messages to get sent to the client out of order. If you do
         // spawn in another thread messages will need to be buffered until the state is sent out
         // on the wire
         let fut = sender.send_message(SubscriptionUpdateMessage { database_update });
```

Comment:
Correct me if I'm wrong, but technically a race condition is possible between 87 and 92.

Reply:
Actually yes, If same client subscribes frequently. Extended the lock to end of this function to fix.

Reply:
Or no, If same client subscribes frequently i.e being serialized by this await - https://github.com/clockworklabs/SpacetimeDB/pull/807/files#diff-cde382f76a7190bac09b3ca01dc55e56abb79225fb34ec657d6e43ba7792da0eR172.

and sending updates together to different clients should not be a problem.

I feel we can do the `drop(subscriptions)` before blocking on send_message. (Being restrictive right now, and not dropping the lock early, until I am sure about this).

@coolreader18 , Would love to get your views on this.